### DISCUSSION TITLE
Inserting cohesive zone and generating mesh from a input polyhedral_mesh.obj file

### DISCUSSION THREAD

**Discussion by Rumi381** [neperfepx/neper#905] [Q&A] [answered]
Hi, everyone. I am new to exploring Neper and facing a particular problem with one of my research. If I have a .obj file containing a polyhedral mesh of a volume, as shown in the figure below, can I leverage Neper, to?
1) load the mesh file,
2) insert cohesive zones at the internal faces of the polyhedral cells
3) mesh the whole model with wedge or tetrahedral elements
4) and Finally, export the resulting mesh as a .inp file to use in Abaqus

I am really struggling with these requirements. Please guide me if anyone knows how to achieve this goal. Here is an example of a polyhedral mesh I will be using:
[3holes_poly.zip](https://github.com/user-attachments/files/16801821/3holes_poly.zip)
![Polyhedral_mesh](https://github.com/user-attachments/assets/3f77cb1c-2371-443c-8d5e-6fe42a8f59ad)

**Comment by rquey** [comment]
It's an interesting case and yes, it should work, but you'll have to write your "polyhedra mesh" as a [Tessellation File (`.tess`)](https://neper.info/doc/fileformat.html#tessellation-file-tess).

May I ask what your model represents? What is your application?

**Reply by Rumi381** [reply]
Thank you, @rquey, for your reply. I am doing a simulation of fracture analysis of Calcification in human Coronary Artery. We will eventually regenerate the 3D model of the arterial section and the Calcification from the OCT or MicroCT image, where the model for the Calcite region will be more complex. But, for simplicity, you consider the artery as a tube and the Calcification inside as a thick, hollow cylinder. So, as per the granular structure and the experimental fracture pattern of the Calcification, we are approximation the grains as polyhedrons, and there will be cohesive zones along the grain boundaries (in 3D cases, along the faces of the polyhedrons). I could achieve the polyhedral structure of the calcite region by leveraging the 'Geogram' package. However, the output file is a .obj file. So, I really need your help to accurately convert the .obj file of the polyhedral mesh to a Neper-supported .tess file format. I want to leverage Neper to insert the cohesive zone along the faces of the polyhedron and also mesh the whole polyhedral mesh with tet or wedge elements since the polyhedrons are just the grains. This is a crucial stage for me since I am doing my simulations in Abaqus. I have never used Neper. Please advise me on how I can accurately and efficiently convert the .obj polyhedral mesh into the .tess file. And please guide me on the subsequent steps I should take, like the appropriate commands in Neper, to achieve my goal. Here is the snapshot of the simplified thick, hollow cylinder polyhedral mesh:
![hollowThickCylinder](https://github.com/user-attachments/assets/e6d9b6a3-6f4b-4cf1-b8cc-fdcf93c01c29)

Another thing I wanted to ask you that If my Polyhedral mesh is in .ovm file format, can we directly use or convert it for Neper?

**Reply by rquey** [reply]
Neper has no import options for CAD models yet, but it could be useful. Can you share a sample obj file? You can attach it here by adding the `.txt` extension.

**Reply by Rumi381** [reply]
Hi, @rquey. I'm glad to hear back from you. I'm unsure if the OpenVolumeMesh (.ovm) file format is considered a CAD model. I asked you about the .ovm file support because this file format has an interesting structure specialized to represent a Polyhedral mesh, and the program I am using is also capable of generating a .ovm file alongside a .obj file. Additionally, I see some closeness between the file structure of .tess and .ovm file format. Below is the standard structure of a .ovm file format I got from [OpenVolumeMesh](https://www.graphics.rwth-aachen.de/media/openvolumemesh_static/Documentation/OpenVolumeMesh-Doc-Latest/file_format.html):

OVM ASCII                   # File header indicating whether data is
                            # coded as plain text or binary
Vertices                    # Indicate that all vertices are listed below
n_v                         # The total number of vertices (>= 0)
x_1 y_1 z_1                 # Coordinate of the first vertex
...
x_n y_n z_n                 # Coordinate of the n_v'th vertex
Edges                       # Indicate that edges are specified below
n_e                         # The total number of edges (>= 0)
vs_1 vt_1                   # First edge's source vertex index followed
...                         # by the first edge's target vertex index  
vs_n vt_n                   # Last edge's source and target vertices
Faces                       # Indicate that faces are specified below
n_f                         # The total number of faces (>= 0)
d he1_1 ... hed_1           # The first face's valence followed by its
...                         # incident half-edges' indices
d he1_n ... hed_n           # Last face's definition
Polyhedra                   # Indicates that polyhedra are specified below
n_c                         # The total number of cells (>= 0)
d hf1_1 ... hfd_1           # The first polyhedron's valence followed
...                         # by its incident half-faces' indices
d hf1_n ... hfd_n           # The last polyhedron's definition

However, the attached zip file contains both the .obj and the .ovm files, which represent the polyhedral mesh of a thick hollow cylinder.
[HollowCylinder_PolyhedralMesh.zip](https://github.com/user-attachments/files/16839323/HollowCylinder_PolyhedralMesh.zip)

Please let me know your findings and how we can leverage Neper to achieve the target.

**Reply by rquey** [reply]
Thanks. I'll look at it.

Meanwhile, you can try to generate a Neper tessellation to mimic your "polyhedral mesh" and mesh it and see if it imports in Abaqus

Without interfacial elements:
```
neper -T -n 100 -reg 1
neper -M n100-id1.tess -for inp
```

With interfacial elements:
```
neper -M n100-id1.tess -interface cohesive -for inp
```

**Reply by rquey** [reply]
@Rumi381 Thanks.  The .ovm format contains indeed more information, but the .obj format is more widely used...

Your obj file does not contain the definition of the polyhedra, although the format allows for it. Here is what Neper outputs as .obj file (which should be consistent with the [obj format specifications](https://en.wikipedia.org/wiki/Wavefront_.obj_file)):
```
$ neper -T -n 2 -format obj
$ more n2-id1.obj

v 1.000000000000 0.000000000000 0.000000000000
v 1.000000000000 1.000000000000 0.000000000000
v 0.000000000000 1.000000000000 0.000000000000
v -0.000000000000 0.072876995892 -0.000000000000
v -0.000000000000 1.000000000000 0.523792986818
v 1.000000000000 1.000000000000 0.921109989703
v 1.000000000000 -0.000000000000 0.356143977106
v 0.103627646137 -0.000000000000 -0.000000000000
v 0.000000000000 0.000000000000 0.000000000000
v 0.000000000000 0.000000000000 1.000000000000
v 0.000000000000 1.000000000000 1.000000000000
v 1.000000000000 1.000000000000 1.000000000000
v 1.000000000000 0.000000000000 1.000000000000
g cell1
f 3 4 5
f 1 2 6 7
f 1 7 8
f 2 3 5 6
f 3 2 1 8 4
f 4 5 6 7 8
g cell2
f 9 10 11 5 4
f 12 13 7 6
f 13 10 9 8 7
f 11 12 6 5
f 9 4 8
f 13 12 11 10
f 4 5 6 7 8
```

It uses `g` to define the polyhedra as groups of faces.

Can you get an obj file with this data?

**Reply by Rumi381** [reply]
Hi @rquey, sorry for the late reply. With *Can you get an obj file with this data?*, did you mean if my input .obj file to *neper* is in the *g cell* form you mentioned, it will be directly compatible with *neper*?

**Reply by rquey** [reply]
The syntax does not have to be exactly the same as shown above, but yes, it should include the `g` fields to define the polyhedra from the faces.

**Reply by Rumi381** [reply]
Hi @rquey. My .obj files only have the vertex and face information. However, there is no information about which faces belong to a certain Polyhedron. I will try to generate one with g fields from the.ovm file. But one I need to be sure of: the .obj format you shared, do the vertices index start from 1 or 0 in the face definition? That means do the vertices follow zero indexing or 1 indexing?

**Reply by rquey** [reply]
1 indexing.

**Reply by rquey** [reply]
This being said, it shouldn't take me too long to add an ovm parser. Let me have a look at it.

**Reply by Rumi381** [reply]
Great! Please let me know your findings.

**Reply by rquey** [reply]
@Rumi381 Your mesh is giant. Do you have something with a (much) smaller number of elements? (both obj and ovm)

**Reply by Rumi381** [reply]
50 Polyhedron is ok? I have one with only 5 Polyhedron.

**Reply by Rumi381** [reply]
Here is one with only 5 cells. Please let me know if you need a little bigger one.
[Block.zip](https://github.com/user-attachments/files/16965240/Block.zip)

**Reply by rquey** [reply]
Seems to work fine from the ovm. I'll make a new version soon.

![d](https://github.com/user-attachments/assets/fbfd5ba4-9bed-4127-baf8-e785710370e8)

PS. Could you share `3holes_poly.ovm`?

**Reply by Rumi381** [reply]
Great! By the way, does Neper currently support an input polyhedral mesh with g fields in a .obj file? If not please also add this one in the new version and let me know when you publish the latest version. That being said, I need your help with the following questions as I am new to using Neper:
1. If I have the input .ovm file containing the polyhedral mesh, what commands should I use to get a very good-quality mesh in Abaqus .inp format (either tetrahedral or hexahedral) with cohesive zones inserted among the faces of the cells? Please let me know all the possible commands.
2. Since we inserted cohesive zones, does the resultant output .inp file include all the necessary things so I can directly open it in Abaqus for my simulations? Or do I have to follow other steps? Please let me know the necessaries.
I eagerly wait for the latest release. Here are the necessary files for the **3holes model**:
[3holes.zip](https://github.com/user-attachments/files/16988085/3holes.zip)

**Reply by rquey** [reply]
It will have support for both obj and ovm, but there are still bugs...

For the rest, see my previous post: https://github.com/neperfepx/neper/discussions/905#discussioncomment-10522662

**Reply by Rumi381** [reply]
Thank you @rque. Would you please briefly comment on the second question regarding cohesive zone insertion?

**Reply by rquey** [reply]
I've updated https://github.com/neperfepx/neper/discussions/905#discussioncomment-10522662 with the correct command.

**Reply by rquey** [reply, marked as answer]
Now available in version 4.10.0.

Typical commands are
```
neper -T -loadtess "file(Block3D_poly.ovm)" -reg 1 -o Block3D_poly
neper -M Block3D_poly.tess -interface cohesive -format inp
```

**Reply by Rumi381** [reply]
Great! I am installing version 4.10.0 but getting the following failed test cases, though I have installed the dependencies. Would you please suggest any quick fix?:

The following tests FAILED:
        147 - M/cl (Failed)
        148 - M/cl_expr (Failed)
        151 - M/clmin (Failed)
        152 - M/clratio (Failed)
        153 - M/cut (Failed)
        155 - M/dim2 (Failed)
        156 - M/dim_expr (Failed)
        158 - M/faset (Failed)
        161 - M/interface (Failed)
        162 - M/interface2 (Failed)
        164 - M/mesh2dalgo_dela (Failed)
        165 - M/mesh2dalgo_fron (Failed)
        166 - M/mesh2dalgo_mead (Failed)
        167 - M/mesh2dalgo_netg (Failed)
        168 - M/mesh3dalgo_netggmne (Failed)
        169 - M/mesh3dalgo_netggmsh (Failed)
        170 - M/mesh3dalgo_netgnetg (Failed)
        171 - M/meshing (Failed)
        173 - M/meshqualdisexpr (Failed)
        174 - M/meshqualexpr (Failed)
        175 - M/meshqualmin (Failed)
        176 - M/morpho_cube (Failed)
        177 - M/msh4 (Failed)
        178 - M/nset1 (Failed)
        179 - M/nset2 (Failed)
        180 - M/nset3 (Failed)
        181 - M/order (Failed)
        188 - M/pl (Failed)
        189 - M/rcl (Failed)
        190 - M/rcl_expr (Failed)
        193 - M/remesh1 (Failed)
        194 - M/remesh2 (Failed)
        195 - M/remesh3 (Failed)
        197 - M/rodrigues_per_cubic (Failed)
        198 - M/rodrigues_per_hexagonal (Failed)
        204 - M/statelsetco (Failed)
        210 - M/stateltco (Failed)
        232 - V/asymptote (Failed)

**Reply by rquey** [reply]
Gmsh not or improperly installed. Same for asymptote.

**Reply by Rumi381** [reply]
I will try installing them again. However, I will test some of my OVM files and get back to you if I have any observations or issues.

**Reply by Rumi381** [reply]
Hi @rquey. I have installed all the dependencies. However, I always get the following error with Module M. Is there any specific reason for this? Please suggest what I should do.

Info   : MODULE  -M loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] n100-id1.tess -for inp
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   :   - Reading arguments...
Info   : Loading input data...
Info   :   - Loading tessellation...
Info   :     [i] Parsing file `n100-id1.tess'...
Info   :     [i] Parsed file `n100-id1.tess'.
Info   : Meshing...
Info   :   - Preparing... (cl = 0.1077) 100%
Info   :   - 0D meshing... 100%
Info   :   - 1D meshing... 100%
Info   :   - 2D meshing... 100% (0.37|0.83/88%| 9%| 3%) 
Info   :   - Fixing 2D-mesh pinches...
Info   :   - 3D meshing... Error  :   - Failed to read msh file
Aborted (core dumped)

**Reply by rquey** [reply]
Are you using version 4.10.1?

**Reply by Rumi381** [reply]
No. I am using version 4.10.0. Is there any issue in 4.10.0?

**Reply by rquey** [reply]
Compatibility issues with the latest versions of Gmsh were fixed in 4.10.1.

**Reply by Zdmire** [reply]
> Great! I am installing version 4.10.0 but getting the following failed test cases, though I have installed the dependencies. Would you please suggest any quick fix?:
> 
> The following tests FAILED: 147 - M/cl (Failed) 148 - M/cl_expr (Failed) 151 - M/clmin (Failed) 152 - M/clratio (Failed) 153 - M/cut (Failed) 155 - M/dim2 (Failed) 156 - M/dim_expr (Failed) 158 - M/faset (Failed) 161 - M/interface (Failed) 162 - M/interface2 (Failed) 164 - M/mesh2dalgo_dela (Failed) 165 - M/mesh2dalgo_fron (Failed) 166 - M/mesh2dalgo_mead (Failed) 167 - M/mesh2dalgo_netg (Failed) 168 - M/mesh3dalgo_netggmne (Failed) 169 - M/mesh3dalgo_netggmsh (Failed) 170 - M/mesh3dalgo_netgnetg (Failed) 171 - M/meshing (Failed) 173 - M/meshqualdisexpr (Failed) 174 - M/meshqualexpr (Failed) 175 - M/meshqualmin (Failed) 176 - M/morpho_cube (Failed) 177 - M/msh4 (Failed) 178 - M/nset1 (Failed) 179 - M/nset2 (Failed) 180 - M/nset3 (Failed) 181 - M/order (Failed) 188 - M/pl (Failed) 189 - M/rcl (Failed) 190 - M/rcl_expr (Failed) 193 - M/remesh1 (Failed) 194 - M/remesh2 (Failed) 195 - M/remesh3 (Failed) 197 - M/rodrigues_per_cubic (Failed) 198 - M/rodrigues_per_hexagonal (Failed) 204 - M/statelsetco (Failed) 210 - M/stateltco (Failed) 232 - V/asymptote (Failed)

Hello,
I encountered the same issue, even after trying different versions of Gmsh and Neper. Unfortunately, the problem with the -M module still persists.
If you have already found a solution to this issue, could you please share how you resolved it? I would really appreciate your help.
Thank you in advance!

**Reply by Rumi381** [reply]
Hi @Zdmire, sorry for the late response. Somehow missed your comment. If you still cannot install Neper, you can follow the instructions in the README.md file I've attached. I hope it will solve the issues.
[README.md](https://github.com/user-attachments/files/21909324/README.md)
